### DISCUSSION TITLE
Unexpected Poor Inference Despite High Evaluation Scores: Validation Loss 0.0031 (Exact: 99.70% | Partial: 99.74%)

### DISCUSSION THREAD

**Discussion by ep0p** [mindee/doctr#1677] [Q&A] [answered]
I have fine-tuned the recognition model crnn_mobilenet_v3_large with a dataset of 3.5 million words, which I have split into three parts: train/val/test. My training command was as follows:
`python references/recognition/train_pytorch.py crnn_mobilenet_v3_large --train_path /home/epop/DATASET/ds_3p5_rec_v1/train --val_path /home/epop/DATASET/ds_3p5_rec_v1/val --name crnn_mobilenet_v3_large_v1 --epochs 10 --batch_size 32 --pretrained --vocab="frenchplus" --device 1 ` 

I then ran the evaluation on my test dataset:
`python references/recognition/evaluate_pytorch.py crnn_mobilenet_v3_large --test_path /home/epop/DATASET/ds_3p5_rec_v1/test --batch_size 32 --vocab="frenchplus" --device 1 --resume crnn_mobilenet_v3_large_v1.pt`

The evaluation yielded a high score:
`Validation loss: 0.00308553 (Exact: 99.70% | Partial: 99.74%)`

However, when running inference, the results are very disappointing, and I am wondering if I missed a step somewhere. Here is the original document:
![Screenshot from 2024-07-24 10-53-50](https://github.com/user-attachments/assets/e55bdab3-662b-4371-8b6c-1a381488438c)

And here is the result of the inference:
![Screenshot from 2024-07-24 10-52-55](https://github.com/user-attachments/assets/6b7ff420-56c0-4d1b-9e4d-3a922fc3c78d)

**Comment by felixdittrich92** [comment]
@ep0p Short question the dataset contains "real" word crops or synth generated ?

**Reply by ep0p** [reply]
Hi @felixdittrich92,

The dataset contains only real word crops, no synthetic data.

I’ve tuned all the recognition models, but they all returned gibberish during inference. 
However, after finishing training parseq with the ```--freeze-backbone``` option i have good results. The recognition is much better now but it still adds a lot of punctuation marks after words.

![Screenshot from 2024-08-20 11-43-23](https://github.com/user-attachments/assets/58eaedb9-d91a-4130-ada3-156475970a38)

**Reply by felixdittrich92** [reply]
Could you share the logs from the run ?

**Reply by ep0p** [reply]
Sure! Here are the logs : [logs_rec.TGZ](https://github.com/user-attachments/files/16674111/logs_rec.TGZ) 
Just for context, the dataset used for these was around 3.5 million real words (an attempt to obtain better results) .
Let me know if you need any additional details.

**Reply by ep0p** [reply]
Hi @felixdittrich92 ,

Do you have any idea why I keep getting random punctuation marks appearing after words during recognition (fine tuned parseq)? 
I’ve cleaned up the images with contrast enhancement, binarization, and denoising, but even after all that, the recognition still returns strange punctuation marks like in these example:

![Screenshot from 2024-08-22 11-18-10](https://github.com/user-attachments/assets/52ac2adf-22f6-4ded-89e0-363b0366d8c8)
![Screenshot from 2024-08-22 11-25-14](https://github.com/user-attachments/assets/40afe79e-b24a-4e32-ac1c-2612b3dee6c4)

**Reply by felixT2K** [reply]
Hi @ep0p 👋🏼 ,
Ok this looks really strange 😅 
Maybe a bias in the feature extractor .. because it was frozen with your latest run.
I would do 2 further experiments with `parseq`. One by training from scratch (without pretrained and freeze_backbone) and one with only pretrained.

**Reply by ep0p** [reply]
Alright, thanks for the suggestion. I’ll try both approaches and will keep you posted!

**Reply by ep0p** [reply]
Hi @felixdittrich92 ,

Since I’m encountering performance issues with detection as well, I’ll add this here, but let me know if you’d prefer me to create a separate post.

I trained two detection models: db_resnet50, db_mobilenet_v3_large  (with and without freeze-backbone), and during inference, they all exhibit the same behavior: detection boxes that are much larger than those of the original models. Here is an example:

1. fine-tuned: db_mobilenet_v3_large

`python references/detection/train_pytorch.py /home/epop/DATASET/ds_1000_det/train /home/epop/DATASET/ds_1000_det/val db_mobilenet_v3_large --name db_mobilenet_v3_large_v1 --wb --epochs 20 --batch_size 32 --freeze-backbone --pretrained --device 1 --lr 0.01 --amp --early-stop --early-stop-epochs 5 --early-stop-delta 0.01
`

![Screenshot from 2024-08-23 09-30-41](https://github.com/user-attachments/assets/ac9ecab2-11fa-4f12-b262-48a269c70794)

2. original: db_mobilenet_v3_large

![Screenshot from 2024-08-23 09-31-57](https://github.com/user-attachments/assets/69230dff-2074-43f9-b11b-000ecc9bae80)

**Reply by felixdittrich92** [reply]
Hi @ep0p :wave:,
Instead of `db_resnet50` i suggest to use `fast_base` instead generalizes much better mostly same inference latency and requires less ressources.
`db_mobilenet_v3_large` is fine as lightweight model.

From my experience it worked best training from scratch if your docs for inference are close to each other (Model does not need to generalize over each thinkable doc format).
With only 300 samples i have trained a well suitable model to predict on docs which are comparable to the train data.
`--freeze-backbone` is not recommended here (for detection training) because the feature extractor will not learn from your new data.

**Reply by ep0p** [reply]
Hi @felixdittrich92 ,

I ran the recognition experiments as suggested:

1. Fine-tuned parseq with only `--pretrained`: The results are  very poor, as shown in the attached graphs. 
`python references/recognition/train_pytorch.py parseq --train_path /home/epop/DATASET/ds_1000_rec_v2/train --val_path /home/epop/DATASET/ds_1000_rec_v2/val --name parseq_optim_v3 --wb --epochs 10 --batch_size 128 --pretrained --vocab="frenchplus" --device 1 --lr 0.01 --early-stop --early-stop-epochs 5 --early-stop-delta 0.01`
![Screenshot from 2024-08-26 08-34-26](https://github.com/user-attachments/assets/c1a5aac6-c4ca-4d7d-b081-5d0866b993e7)



3. Parseq trained from scratch: The performance is similarly poor
`python references/recognition/train_pytorch.py parseq --train_path /home/epop/DATASET/ds_1000_rec_v2/train --val_path /home/epop/DATASET/ds_1000_rec_v2/val --name parseq_optim_v_train_from_scratch --wb --epochs 10 --batch_size 128 --vocab="frenchplus" --device 1 --lr 0.01 --early-stop --early-stop-epochs 5 --early-stop-delta 0.01`
![Screenshot from 2024-08-26 08-35-42](https://github.com/user-attachments/assets/487189bd-700b-4915-b47d-1b74d84daa06)


Both versions yield poor inference results. 
The only improvement I see is when I use the `--freeze-backbone` option. However, as I mentioned earlier, this approach introduces too many unnecessary punctuation marks, which degrades the output quality.

**Reply by felixdittrich92** [reply, marked as answer]
Hi @ep0p :wave:,

What i see directly is `--lr 0.01` what was the reason to use such a high value ? :sweat_smile:

Mh.. we pretrained the models `with the default args (only epochs increased to 20 - so maybe with less data set to 40 or 50)` from scratch on ~10M samples (3M are definitely enough especially for fine tuning).

I stuck a bit because i have fine tuned `parseq` also (https://huggingface.co/Felix92/doctr-torch-parseq-multilingual-v1) only on ~1M synth samples without trouble. (with `--pretrained` and `50 epochs` other args unchanged)

**Reply by ep0p** [reply]
You’re right, I chose the learning rate based on the find-lr graph.
![Screenshot from 2024-08-26 09-18-00](https://github.com/user-attachments/assets/e8f12b8f-fd12-4810-a170-a6a257c74c44)

I could have gone with something between 0.001 and 0.01, but I went straight to the upper limit 🙈—like an idiot!
Thanks for the insight on the pretraining setup.

I’ve just started another fine-tuning process with the default LR. This time, I’m hoping for better results. 🤞

**Reply by felixdittrich92** [reply]
Let's hope :sweat_smile:

**Reply by ep0p** [reply]
Hi @felixdittrich92 ,

I wanted to share the results from my latest parseq fine-tuning experiment:

Dataset: 3 million words
Vocab : 'frenchplus' (`VOCABS["frenchplus"] = VOCABS["french"] + "»«m²§•·–’œ≥◊×≤½æ‰‘Œ♦µ/βγ¹”º…öð—¤’¼¦äíáå¨"`) 
Other parameters: Default settings
Epochs: 20

The model achieved high scores comparable to the previous runs.
![image](https://github.com/user-attachments/assets/3f39add9-b3ca-4a3b-8879-e0f4dbe01186)

But unfortunately, the inference results are still poor.
![Screenshot from 2024-08-28 09-48-01](https://github.com/user-attachments/assets/6485a4f7-6aac-4593-9010-6b76d890f6d7)

I tried the model you provided, and it works perfectly. 
At this point, I'm wondering if the issue might be related to the vocabulary.

**Reply by felixdittrich92** [reply]
@ep0p I have a guess, could you please share the code you use for the inference?

**Reply by ep0p** [reply]
Sure, here’s what I’m using : 
```
doc = DocumentFile.from_pdf("/home/elenapop/Downloads/108755fre.pdf")

#reco_model_from_hub = from_hub('Felix92/doctr-torch-parseq-multilingual-v1')
reco_model = parseq(pretrained=True, vocab=VOCABS["frenchplus"])
reco_params = torch.load('/home/elenapop/Downloads/parseq_optim_v4.pt', map_location="cpu")
reco_model.load_state_dict(reco_params)

model = ocr_predictor(det_arch='db_mobilenet_v3_large', reco_arch=reco_model, pretrained=True)
model.det_predictor.model.postprocessor.bin_thresh = 0.3


result = model(doc)
print(result.render())
```

**Reply by felixdittrich92** [reply]
> Sure, here’s what I’m using :
> 
> ```
> doc = DocumentFile.from_pdf("/home/elenapop/Downloads/108755fre.pdf")
> 
> #reco_model_from_hub = from_hub('Felix92/doctr-torch-parseq-multilingual-v1')
> reco_model = parseq(pretrained=True, vocab=VOCABS["frenchplus"])
> reco_params = torch.load('/home/elenapop/Downloads/parseq_optim_v4.pt', map_location="cpu")
> reco_model.load_state_dict(reco_params)
> 
> model = ocr_predictor(det_arch='db_mobilenet_v3_large', reco_arch=reco_model, pretrained=True)
> model.det_predictor.model.postprocessor.bin_thresh = 0.3
> 
> 
> result = model(doc)
> print(result.render())
> ```

Mh no ..everything is correct :+1: 
Unfortunately i'm really not sure how i can further help .. I'm slowly running out of ideas :sweat_smile: 

Especially the extensive prediction of punctuations is not really clear is not clear to me ..

Does it predict the "special" symbols in any case correct ? Or does it predict a dot instead ?

**Reply by felixdittrich92** [reply]
Could you create a char count map from your dataset to get a better understanding ?
Something like:

```
{"a": 127, "b": 234, ..}
```

**Reply by ep0p** [reply]
I might be onto something ... my json looks like this 
```
   "word_83074.jpg": "Mme",
    "word_1871413.jpg": "crise",
    "word_1215541.jpg": "qu'il",
    "word_2850502.jpg": "\u00e0",
    "word_2110575.jpg": "III",
    "word_1377404.jpg": "le",
    "word_2505949.jpg": "travail,",
    "word_3300455.jpg": "comp\u00e9tences",
    "word_859743.jpg": "DE",
    "word_19140.jpg": ";",
    "word_3110300.jpg": "la",
    "word_362965.jpg": "\u00e0",
    "word_674002.jpg": "par",
    "word_2910512.jpg": "\u00e0",
```
 When splitting into train/val, I have lost the encoding  `(coding='utf-8', ensure_ascii=False)`

**Reply by felixdittrich92** [reply]
> I might be onto something ... my json looks like this
> 
> ```
>    "word_83074.jpg": "Mme",
>     "word_1871413.jpg": "crise",
>     "word_1215541.jpg": "qu'il",
>     "word_2850502.jpg": "\u00e0",
>     "word_2110575.jpg": "III",
>     "word_1377404.jpg": "le",
>     "word_2505949.jpg": "travail,",
>     "word_3300455.jpg": "comp\u00e9tences",
>     "word_859743.jpg": "DE",
>     "word_19140.jpg": ";",
>     "word_3110300.jpg": "la",
>     "word_362965.jpg": "\u00e0",
>     "word_674002.jpg": "par",
>     "word_2910512.jpg": "\u00e0",
> ```
> 
> When splitting into train/val, I have lost the encoding `(coding='utf-8', ensure_ascii=False)`

Ups..this is a problem  ...maybe already the explanation

**Reply by ep0p** [reply]
Yeah.... i guess i'll go and cry in a corner now ....

**Reply by felixdittrich92** [reply]
> Yeah.... i guess i'll go and cry in a corner now ....


Can happen. It's annoying though, I agree

**Comment by ep0p** [comment]
Hi @felixdittrich92 ,

After fixing my dataset and fine-tuning parseq again, I'm still not getting good results (words are messed up, and there are a lot of punctuation marks):
![image](https://github.com/user-attachments/assets/b35158fd-7e67-4037-80a4-058c1d0f8d4e)


 `python references/recognition/train_pytorch.py parseq --train_path /home/epop/DATASET/ds_3mil_rec/train --val_path /home/epop/DATASET/ds_3mil_rec/val --name parseq_newDS_P --wb --epochs 20 --batch_size 128 --pretrained --vocab="frenchplus" --device 1`

I have also created a character map for my dataset:
[char_count_map.json](https://github.com/user-attachments/files/16845965/char_count_map.json)

The only custom parameter I've used is the vocab. 
I believe 20 epochs should be sufficient for a dataset of 3 million words, I don't see any other mistake that i may be doing 

I was wondering if there is a cache or something of the models and maybe i am tuning a bad model (that was badly tuned before) ... or some other crazy explanation of why this is not working for me.

My next, and last step in all this problem would be creating a synthetic ds to test.
If possible, could you share the namespace log of the model that you provided (https://huggingface.co/Felix92/doctr-torch-parseq-multilingual-v1) ?

Edit: Or can it be because of the size of the image in the dataset? here is an example: 
![word_1000049](https://github.com/user-attachments/assets/e0fd9991-18c2-4762-8fdb-0940be65d55f)

At this stage i am doubting everything ...

**Reply by felixdittrich92** [reply]
Hey @ep0p :wave:,

The models are cached in `~/.cache/doctr/models`  but we trained only 1 version of parseq :)

Args i used:
```
"args": [
        "parseq",
        "--train_path=/home/felix/Desktop/synthtiger/800k_multilingual_v2_train",
        "--val_path=/home/felix/Desktop/synthtiger/200k_multilingual_v2_val",
        "-b",
        "128",
        "--pretrained",
        "--vocab=multilingual",
        "--epochs=50",
    ],
```

About the word crop size it doesn't matter because it's resized to 32x128 by keeping aspect ratio.
